internal/generator/fri/template/fri.go.tmpl

```diff
@@ -0,0 +1,700 @@
+
+	for i := 0; i < s/2; i++ {
+
+		p1.Add(&pSorted[2*i], &pSorted[2*i+1])
+		p2.Sub(&pSorted[2*i], &pSorted[2*i+1]).Mul(&p2, &acc)
+		res[i].Mul(&p2, &x).Add(&res[i], &p1).Mul(&res[i], &twoInv)
+
+		acc.Mul(&acc, &gInv)
+
+	}
+
+	return res
+}
+
+// buildProofOfProximitySingleRound generates a proof that a function, given as an oracle from
+// the verifier point of view, is in fact δ-close to a polynomial.
+// * salt is a variable for multi rounds, it allows to generate different challenges using Fiat Shamir
+// * p is in evaluation form
+func (s radixTwoFri) buildProofOfProximitySingleRound(salt fr.Element, p []fr.Element) (round, error) {
+
+	// the proof will contain nbSteps interactions
+	var res round
+	res.interactions = make([][2]partialMerkleProof, s.nbSteps)
+
+	// Fiat Shamir transcript to derive the challenges. The xᵢ are used to fold the
+	// polynomials.
+	// During the i-th round, the prover has a polynomial P of degree n. The verifier sends
+	// xᵢ∈ Fᵣ to the prover. The prover expresses F in Fᵣ[X,Y]/<Y-X²> as
+	// P₀(Y)+X P₁(Y) where P₀, P₁ are of degree n/2, and he then folds the polynomial
+	// by replacing x by xᵢ.
+	xis := make([]string, s.nbSteps+1)
+	for i := 0; i < s.nbSteps; i++ {
+		xis[i] = fmt.Sprintf("x%d", i)
+	}
+	xis[s.nbSteps] = "s0"
+	fs := fiatshamir.NewTranscript(s.h, xis...)
+
+	// the salt is binded to the first challenge, to ensure the challenges
+	// are different at each round.
+	err := fs.Bind(xis[0], salt.Marshal())
+	if err != nil {
+		return round{}, err
+	}
+
+	// step 1 : fold the polynomial using the xi
+
+	// evalsAtRound stores the list of the nbSteps polynomial evaluations, each evaluation
+	// corresponds to the evaluation o the folded polynomial at round i.
+	evalsAtRound := make([][]fr.Element, s.nbSteps)
+
+	// evaluate p and sort the result
+	_p := make([]fr.Element, s.domain.Cardinality)
+	copy(_p, p)
+
+	// gInv inverse of the generator of the cyclic group of size the size of the polynomial.
+	// The size of the cyclic group is ρ*s.domainSize, and not s.domainSize.
+	var gInv fr.Element
+	gInv.Set(&s.domain.GeneratorInv)
+
+	for i := 0; i < s.nbSteps; i++ {
+
+		evalsAtRound[i] = sort(_p)
+
+		// compute the root hash, needed to derive xi
+		t := merkletree.New(s.h)
+		for k := 0; k < len(_p); k++ {
+			t.Push(evalsAtRound[i][k].Marshal())
+		}
+		rh := t.Root()
+		err := fs.Bind(xis[i], rh)
+		if err != nil {
+			return res, err
+		}
+
+		// derive the challenge
+		bxi, err := fs.ComputeChallenge(xis[i])
+		if err != nil {
+			return res, err
+		}
+		var xi fr.Element
+		xi.SetBytes(bxi)
+
+		// fold _p, reusing its memory
+		_p = foldPolynomialLagrangeBasis(evalsAtRound[i], gInv, xi)
+
+		// g <- g²
+		gInv.Square(&gInv)
+
+	}
+
+	// last round, provide the evaluation. The fully folded polynomial is of size rho. It should
+	// correspond to the evaluation of a polynomial of degree 1 on ρ points, so those points
+	// are supposed to be on a line.
+	res.evaluation = make([]fr.Element, rho)
```

Comment:
if `rho` is a constant, `evaluation` should be an array of size rho in the struct, not a slice